pyexchange/uniswapv2.py

```diff
@@ -0,0 +1,342 @@
+# This file is part of Maker Keeper Framework.
+#
+# Copyright (C) 2020 MikeHathaway
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import time
+from web3 import Web3
+from typing import List
+
+from pymaker import Contract, Address, Transact, Wad
+from pymaker.token import ERC20Token
+from pymaker.model import Token
+from pymaker.approval import directly
+from pyexchange.graph import GraphClient
+from pyexchange.model import Trade
+
+
+class UniswapV2(Contract):
+    """
+    UniswapV2 Python Client
+
+    Each UniswapV2 instance is intended to be used with a single pool at a time.
+
+    Documentation is available here: https://uniswap.org/docs/v2/
+    """
+
+    pair_abi = Contract._load_abi(__name__, 'abi/IUniswapV2Pair.abi')
+    router_abi = Contract._load_abi(__name__, 'abi/IUniswapV2Router02.abi')
+    factory_abi = Contract._load_abi(__name__, 'abi/IUniswapV2Factory.abi')
+    router_bin = Contract._load_bin(__name__, 'abi/IUniswapV2Router02.bin')
+    factory_bin = Contract._load_bin(__name__, 'abi/IUniswapV2Factory.bin')
+
+    def __init__(self, web3: Web3, token_a: Token, token_b: Token):
+                 # ec_signature_r: Optional[str], ec_signature_s: Optional[str], ec_signature_v: Optional[int]):
+        assert (isinstance(web3, Web3))
+        assert (isinstance(token_a, Token))
+        assert (isinstance(token_b, Token))
+
+        self.web3 = web3
+        self.token_a = token_a
+        self.token_b = token_b
+        self.router_address = Address("0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D")
+        self.factory_address = Address("0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f")
```

Comment:
Hardcoding these addresses prevents testing outside of _mainnet_.  Also the **router** address is not immutable; that contract can be swapped out because no balances are stored there.

Reply:
The address is the same across all networks, but that being said, I agree hardcoding is non ideal. Do you have a preferred solution to this? I've seen both Enums and address config files in the rest of the codebase.

Reply:
I'm happy keeping them as ctor arguments with default values, as you've implemented.  Keepers can override them with commandline arguments if desired.

Reply:
Left as is, per discussion.